File: app/src/main/java/org/mifos/mobilebanking/ui/fragments/SavingsAccountWithdrawFragment.java
package org.mifos.mobilebanking.ui.fragments;

/*
 * Created by saksham on 02/July/2018
 */

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputLayout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.models.accounts.savings.SavingsAccountWithdrawPayload;
import org.mifos.mobilebanking.models.accounts.savings.SavingsWithAssociations;
import org.mifos.mobilebanking.presenters.SavingsAccountWithdrawPresenter;
import org.mifos.mobilebanking.ui.activities.base.BaseActivity;
import org.mifos.mobilebanking.ui.fragments.base.BaseFragment;
import org.mifos.mobilebanking.ui.views.SavingsAccountWithdrawView;
import org.mifos.mobilebanking.utils.Constants;
import org.mifos.mobilebanking.utils.MFDatePicker;
import org.mifos.mobilebanking.utils.Toaster;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class SavingsAccountWithdrawFragment extends BaseFragment
        implements SavingsAccountWithdrawView {

    @BindView(R.id.tv_client_name)
    TextView tvClientName;

    @BindView(R.id.tv_account_number)
    TextView tvAccountNumber;

    @BindView(R.id.tv_withdrawal_date)
    TextView tvWithdrawalDate;

    @BindView(R.id.til_remark)
    TextInputLayout tilRemark;

    @Inject
    SavingsAccountWithdrawPresenter presenter;

    private View rootView;
    private SavingsWithAssociations savingsWithAssociations;
    private SavingsAccountWithdrawPayload payload;

    public static SavingsAccountWithdrawFragment newInstance(
            SavingsWithAssociations savingsWithAssociations) {
        SavingsAccountWithdrawFragment fragment = new SavingsAccountWithdrawFragment();

        Bundle bundle = new Bundle();
        bundle.putParcelable(Constants.SAVINGS_ACCOUNTS, savingsWithAssociations);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            savingsWithAssociations = getArguments().getParcelable(Constants.SAVINGS_ACCOUNTS);
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_savings_account_withdraw_fragment,
                container, false);
        ButterKnife.bind(this, rootView);
        ((BaseActivity) getActivity()).getActivityComponent().inject(this);
        presenter.attachView(this);
        showUserInterface();
        return rootView;
    }

    @OnClick(R.id.btn_withdraw_savings_account)
    void onWithdrawSavingsAccount() {
        submitWithdrawSavingsAccount();
    }

    @Override
    public void showUserInterface() {
        getActivity().setTitle(getString(R.string.withdraw_savings_account));
        tvAccountNumber.setText(savingsWithAssociations.getAccountNo());
        tvClientName.setText(savingsWithAssociations.getClientName());
        tvWithdrawalDate.setText(MFDatePicker.getDatePickedAsString());
    }

    public boolean isFormIncomplete() {
        boolean rv = false;
        if (tilRemark.getEditText().getText().toString().trim().length() == 0) {
            rv = true;
            tilRemark.setError(getString(R.string.error_validation_blank,
                    getString(R.string.remark)));
        }
        return rv;
    }

    public void submitWithdrawSavingsAccount() {
        tilRemark.setErrorEnabled(false);
        if (!isFormIncomplete()) {
            payload = new SavingsAccountWithdrawPayload();
            payload.setNote(tilRemark.getEditText().getText().toString());
            payload.setWithdrawnOnDate(MFDatePicker.getDatePickedAsString());
            presenter.submitWithdrawSavingsAccount(savingsWithAssociations.getAccountNo(), payload);
        }
    }

    @Override
    public void showSavingsAccountWithdrawSuccessfully() {
        showMessage(getString(R.string.savings_account_withdraw_successful));
        getActivity().getSupportFragmentManager().popBackStack();
    }

    @Override
    public void showMessage(String message) {
        Toaster.show(rootView, message);
    }

    @Override
    public void showError(String error) {
        Toaster.show(rootView, error);
    }

    @Override
    public void showProgress() {
        showMifosProgressDialog(getString(R.string.progress_message_loading));

    }

    @Override
    public void hideProgress() {
        hideMifosProgressDialog();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.detachView();
    }
}

File: app/src/main/java/org/mifos/mobile/ui/fragments/SavingsAccountWithdrawFragment.kt
package org.mifos.mobile.ui.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import org.mifos.mobile.R
import org.mifos.mobile.databinding.FragmentSavingsAccountWithdrawFragmentBinding
import org.mifos.mobile.models.accounts.savings.SavingsAccountWithdrawPayload
import org.mifos.mobile.models.accounts.savings.SavingsWithAssociations
import org.mifos.mobile.ui.activities.SavingsAccountContainerActivity
import org.mifos.mobile.ui.fragments.base.BaseFragment
import org.mifos.mobile.utils.Constants
import org.mifos.mobile.utils.ParcelableAndSerializableUtils.getCheckedParcelable
import org.mifos.mobile.utils.SavingsAccountUiState
import org.mifos.mobile.utils.Toaster
import org.mifos.mobile.utils.getTodayFormatted
import org.mifos.mobile.viewModels.SavingsAccountWithdrawViewModel

/*
* Created by saksham on 02/July/2018
*/
@AndroidEntryPoint
class SavingsAccountWithdrawFragment : BaseFragment() {

    private var _binding: FragmentSavingsAccountWithdrawFragmentBinding? = null
    private val binding get() = _binding!!
    private val viewModel: SavingsAccountWithdrawViewModel by viewModels()
    private var savingsWithAssociations: SavingsWithAssociations? = null
    private var payload: SavingsAccountWithdrawPayload? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (activity as? SavingsAccountContainerActivity)?.showToolbar()
        if (arguments != null) {
            savingsWithAssociations = arguments?.getCheckedParcelable(
                SavingsWithAssociations::class.java,
                Constants.SAVINGS_ACCOUNTS
            )        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        _binding = FragmentSavingsAccountWithdrawFragmentBinding.inflate(inflater, container, false)
        showUserInterface()
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.btnWithdrawSavingsAccount.setOnClickListener {
            onWithdrawSavingsAccount()
        }

        viewLifecycleOwner.lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.savingsAccountWithdrawUiState.collect { state ->
                    when (state) {
                        SavingsAccountUiState.Loading -> showProgress()

                        is SavingsAccountUiState.ErrorMessage -> {
                            hideProgress()
                            showError(state.error.message ?: "")
                        }

                        SavingsAccountUiState.SavingsAccountWithdrawSuccess -> {
                            hideProgress()
                            showSavingsAccountWithdrawSuccessfully()
                        }

                        is SavingsAccountUiState.Initial -> {}

                        else -> throw IllegalStateException("Unexpected state : $state")
                    }
                }
            }
        }
    }

    private fun onWithdrawSavingsAccount() {
        submitWithdrawSavingsAccount()
    }

    fun showUserInterface() {
        activity?.title = getString(R.string.withdraw_savings_account)
        binding.tvAccountNumber.text = savingsWithAssociations?.accountNo
        binding.tvClientName.text = savingsWithAssociations?.clientName
        binding.tvWithdrawalDate.text = getTodayFormatted()
    }

    private val isFormIncomplete: Boolean
        get() {
            var rv = false
            if (binding.tilRemark.editText?.text.toString().trim { it <= ' ' }.isEmpty()) {
                rv = true
                binding.tilRemark.error = getString(
                    R.string.error_validation_blank,
                    getString(R.string.remark),
                )
            }
            return rv
        }

    private fun submitWithdrawSavingsAccount() {
        binding.tilRemark.isErrorEnabled = false
        if (!isFormIncomplete) {
            payload = SavingsAccountWithdrawPayload()
            payload?.note = binding.tilRemark.editText?.text.toString()
            payload?.withdrawnOnDate = getTodayFormatted()
            viewModel.submitWithdrawSavingsAccount(savingsWithAssociations?.accountNo, payload)
        }
    }

    private fun showSavingsAccountWithdrawSuccessfully() {
        showMessage(getString(R.string.savings_account_withdraw_successful))
        activity?.supportFragmentManager?.popBackStack()
    }

    private fun showMessage(message: String?) {
        Toaster.show(binding.root, message)
    }

    fun showError(error: String?) {
        Toaster.show(binding.root, error)
    }

    fun showProgress() {
        showMifosProgressDialog(getString(R.string.progress_message_loading))
    }

    fun hideProgress() {
        hideMifosProgressDialog()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        fun newInstance(
            savingsWithAssociations: SavingsWithAssociations?,
        ): SavingsAccountWithdrawFragment {
            val fragment = SavingsAccountWithdrawFragment()
            val bundle = Bundle()
            bundle.putParcelable(Constants.SAVINGS_ACCOUNTS, savingsWithAssociations)
            fragment.arguments = bundle
            return fragment
        }
    }
}
